### DISCUSSION TITLE
recon_loss negative samples randomly generated

### DISCUSSION THREAD

**Discussion by TommyGiak** [pyg-team/pytorch_geometric#7476] [General]
Hi everyone,
is not clear to me why have to generate negative random samples each call of recon_loss().
I mean here:
```python
def recon_loss(self, z: Tensor, pos_edge_index: Tensor, neg_edge_index: Optional[Tensor] = None) -> Tensor:
  ...
  if neg_edge_index is None:
    neg_edge_index = negative_sampling(pos_edge_index, z.size(0))
  ...
```
My doubts are about real data (i.e. protein-protein interaction), we only have a fixed number of negative samples and maybe most of the links (edges) are uncertain or probable/unprobable. 
Using only the fixed and usually few number of negative samples can lead to a overfitting and low precision (right?) so how should we behave?
Is it ok to randomly create negative samples also in cases like those to avoid overfitting?
If you can also suggest some detailed material where to study this kind of problem it would be appreciated.
Thanks!

**Comment by rusty1s** [comment]
I don't have a good pointer to you, and it certainly depends on your problem. In most cases, I would say sampling negatives on-the-fly yields the best result, but this is ofc not necessarily the case in which non-positive edges are not necessarily negative (or at least not negative most of the time).
